File: verify_xirr.js
// 复刻 定投测算（净值）.html 的目标止盈定投现金流，并输出供 Excel 核对
const fs = require('fs');

// ---------- 参数（请按你当时的设置修改） ----------
const PARAMS = {
  fundContains: '008114',     // 基金筛选
  strategy: '5',              // 目标止盈定投
  freq: 'monthly',
  dayOfMonth: '1',            // 每月1号
  baseAmount: 1000,
  stopGainPct: 8,             // 止盈阈值 8%
  stopGainSellRatio: 100,     // 赎回比例 100%
  startDate: '2019-12-10',
  endDate: '2026-07-07',
  div: 'reinvest'             // 红利再投资
};

// ---------- 工具函数（与代码一致） ----------
function formatDate(d) {
  const y = d.getUTCFullYear();
  const m = String(d.getUTCMonth() + 1).padStart(2, '0');
  const day = String(d.getUTCDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}
function xirr(cashFlows, dates, guess = 0.1) {
  if (cashFlows.length !== dates.length || cashFlows.length < 2) return NaN;
  const paired = cashFlows.map((cf, i) => ({ cf, date: dates[i] }));
  paired.sort((a, b) => a.date - b.date);
  const sortedFlows = paired.map(p => p.cf);
  const sortedDates = paired.map(p => p.date);
  const baseDate = sortedDates[0];
  const dayFrac = j => (sortedDates[j] - baseDate) / (1000 * 60 * 60 * 24) / 365;
  const npv = (rate) => {
    let s = 0;
    for (let j = 0; j < sortedFlows.length; j++) s += sortedFlows[j] / Math.pow(1 + rate, dayFrac(j));
    return s;
  };
  const tolerance = 1e-7, maxIter = 100;
  let rate = guess;
  for (let i = 0; i < maxIter; i++) {
    let v = 0, d = 0;
    for (let j = 0; j < sortedFlows.length; j++) {
      const f = dayFrac(j);
      const term = Math.pow(1 + rate, f);
      v += sortedFlows[j] / term;
      d -= sortedFlows[j] * f * Math.pow(1 + rate, f - 1);
    }
    if (Math.abs(v) < tolerance) return rate;
    if (Math.abs(d) < tolerance) break;
    rate -= v / d;
  }
  let lo = -0.9999, hi = 100;
  let fLo = npv(lo), fHi = npv(hi);
  if (fLo * fHi > 0) return NaN;
  for (let i = 0; i < 200; i++) {
    const mid = (lo + hi) / 2;
    const fMid = npv(mid);
    if (Math.abs(fMid) < tolerance) return mid;
    if (fLo * fMid < 0) { hi = mid; fHi = fMid; }
    else { lo = mid; fLo = fMid; }
  }
  return (lo + hi) / 2;
}

// ---------- 读取基金数据 ----------
const raw = JSON.parse(fs.readFileSync('回测项目_2026-07-09.json', 'utf8'));
const fundsData = raw.fundsData;
const fkey = Object.keys(fundsData).find(k => k.includes(PARAMS.fundContains));
if (!fkey) { console.error('未找到基金:', PARAMS.fundContains); process.exit(1); }
const fund = fundsData[fkey];
console.log('基金:', fkey, ' 交易日数:', fund.dates.length);

const startTs = new Date(PARAMS.startDate + 'T00:00:00').getTime();
const endTs = new Date(PARAMS.endDate + 'T00:00:00').getTime();

const dates = [], dateStrs = [], navs = [], divs = [];
for (let i = 0; i < fund.dates.length; i++) {
  const ts = new Date(fund.dates[i] + 'T00:00:00').getTime();
  if (ts >= startTs && ts <= endTs) {
    dates.push(new Date(fund.dates[i] + 'T00:00:00'));
    dateStrs.push(fund.dates[i]);
    navs.push(fund.nav[i]);
    divs.push(fund.div[i] || 0);
  }
}
const N = dates.length;
console.log('区间交易日数:', N);

const dow = new Array(N), dom = new Array(N);
for (let k = 0; k < N; k++) { dow[k] = dates[k].getUTCDay(); dom[k] = parseInt(dateStrs[k].split('-')[2], 10); }

const reinvest = PARAMS.div === 'reinvest';
const strat = PARAMS.strategy;
const isInvest = new Array(N).fill(false);
if (strat === '1') isInvest[0] = true;
else {
  const freq = PARAMS.freq;
  if (freq === 'weekly' || freq === 'biweekly') {
    const wd = parseInt(PARAMS.weekday || '1');
    let firstIdx = -1;
    for (let k = 0; k < N; k++) { if (dow[k] === wd) { firstIdx = k; break; } }
    if (firstIdx >= 0) for (let k = firstIdx; k < N; k++) {
      if (dow[k] === wd) {
        if (freq === 'biweekly') { const wd2 = Math.round((dates[k] - dates[firstIdx]) / (7 * 86400000)); if (wd2 % 2 !== 0) continue; }
        isInvest[k] = true;
      }
    }
  } else {
    if (PARAMS.dayOfMonth === 'first') {
      let lastMonth = -1;
      for (let k = 0; k < N; k++) { const ym = dateStrs[k].slice(0, 7); if (ym !== lastMonth) { isInvest[k] = true; lastMonth = ym; } }
    } else {
      const domN = parseInt(PARAMS.dayOfMonth);
      for (let k = 0; k < N; k++) if (dom[k] === domN) isInvest[k] = true;
    }
  }
}
const investIdx = [];
for (let k = 0; k < N; k++) if (isInvest[k]) investIdx.push(k);
console.log('定投次数:', investIdx.length);

let shares = 0, totalCash = 0, totalInvested = 0, costBasis = 0, stopGainCount = 0, investCount = 0, totalRedeemed = 0, totalDividendCash = 0;
let runPrincipal = 0, maxPrincipal = 0, runMaxPrincipal = 0;
const cashFlows = [], flowDates = [];

for (let k = 0; k < N; k++) {
  const nav = navs[k];
  const date = dates[k];
  if (shares > 0 && divs[k] > 0) {
    const totalDiv = shares * divs[k];
    if (reinvest) shares += totalDiv / nav;
    else { totalCash += totalDiv; totalDividendCash += totalDiv; cashFlows.push(totalDiv); flowDates.push(date); }
  }
  let amt = 0;
  if (isInvest[k]) {
    if (strat === '1') amt = PARAMS.baseAmount;
    else if (strat === '2') amt = PARAMS.baseAmount;
    else if (strat === '3') { investCount++; amt = investCount * PARAMS.baseAmount - shares * nav; }
    else if (strat === '4') amt = PARAMS.baseAmount;
    else if (strat === '5') amt = PARAMS.baseAmount;
    else if (strat === '6') amt = PARAMS.baseAmount;
  }
  if (strat === '5') {
    const th = parseFloat(PARAMS.stopGainPct) / 100;
    const sellRatio = Math.min(1, Math.max(0, parseFloat(PARAMS.stopGainSellRatio) / 100));
    const roundPrincipal = runMaxPrincipal;
    if (roundPrincipal > 0 && sellRatio > 0) {
      const holdMv = shares * nav;
      if ((holdMv - costBasis) / roundPrincipal >= th) {
        const sellShares = shares * sellRatio;
        const proceeds = sellShares * nav;
        shares -= sellShares; totalCash += proceeds; totalRedeemed += proceeds;
        cashFlows.push(proceeds); flowDates.push(date);
        costBasis *= (1 - sellRatio);
        stopGainCount++;
        if (sellRatio >= 1 || shares < 1e-9) { runPrincipal = 0; runMaxPrincipal = 0; }
      }
    }
  }
  if (amt !== 0 && isInvest[k]) {
    if (amt > 0) {
      shares += amt / nav; totalInvested += amt; costBasis += amt; runPrincipal += amt;
      cashFlows.push(-amt); flowDates.push(date);
    } else {
      let sellShares = (-amt) / nav;
      if (sellShares > shares) sellShares = shares;
      if (sellShares > 0) {
        const proceeds = sellShares * nav;
        shares -= sellShares; totalCash += proceeds; totalRedeemed += proceeds;
        cashFlows.push(proceeds); flowDates.push(date);
        const denom = shares + sellShares;
        if (denom > 0) costBasis *= shares / denom;
      }
    }
  }
  runMaxPrincipal = Math.max(runMaxPrincipal, runPrincipal);
  maxPrincipal = Math.max(maxPrincipal, runPrincipal);
}

const finalDate = dates[N - 1];
cashFlows.push(shares * navs[N - 1]); flowDates.push(finalDate);

const xirrVal = xirr(cashFlows, flowDates) * 100;
const finalAsset = shares * navs[N - 1] + totalCash;

// ---------- 输出摘要 ----------
console.log('\n===== 摘要 =====');
console.log('总投入本金 totalInvested =', totalInvested.toFixed(0));
console.log('期末总资产 finalAsset   =', finalAsset.toFixed(0));
console.log('累积赎回 totalRedeemed  =', totalRedeemed.toFixed(0));
console.log('累计现金分红            =', totalDividendCash.toFixed(0));
console.log('止盈次数 stopGainCount  =', stopGainCount);
console.log('最大投入本金 maxPrincipal =', maxPrincipal.toFixed(0));
console.log('期末剩余份额市值        =', (shares * navs[N-1]).toFixed(0));
console.log('代码 xirr(牛顿guess=0.1) =', (isNaN(xirrVal)?'NaN':xirrVal.toFixed(4)+'%'));

// ---------- 扫描所有实根（检测伪根） ----------
function npvRate(rate) {
  const paired = cashFlows.map((cf, i) => ({ cf, date: flowDates[i] }));
  paired.sort((a, b) => a.date - b.date);
  const sf = paired.map(p => p.cf), sd = paired.map(p => p.date);
  const bd = sd[0];
  let s = 0;
  for (let j = 0; j < sf.length; j++) s += sf[j] / Math.pow(1 + rate, (sd[j] - bd) / 86400000 / 365);
  return s;
}
const roots = [];
let prevR = -0.9999, prevV = npvRate(prevR);
for (let r = -0.9999 + 0.0001; r <= 100; r += 0.0001) {
  const v = npvRate(r);
  if (prevV === 0) { /* */ }
  if (prevV * v < 0) {
    // 细化
    let lo = prevR, hi = r;
    for (let it = 0; it < 60; it++) { const mid = (lo+hi)/2; if (npvRate(lo)*npvRate(mid) <= 0) hi = mid; else lo = mid; }
    roots.push((lo+hi)/2);
  }
  prevR = r; prevV = v;
}
console.log('\n===== 所有实根 (NPV=0 的 r) =====');
if (roots.length === 0) console.log('无实根 (区间内 NPV 不变号)');
else roots.forEach((r, i) => console.log(`根${i+1}: ${(r*100).toFixed(4)}%`));

// ---------- 输出现金流表（供 Excel 粘贴） ----------
console.log('\n===== 现金流表 (日期<TAB>金额) 复制粘贴到 Excel 两列 =====');
// 按日期排序便于核对
const paired = cashFlows.map((cf, i) => ({ d: flowDates[i], cf }));
paired.sort((a, b) => a.d - b.d);
let out = '';
paired.forEach(p => { out += formatDate(p.d) + '\t' + p.cf.toFixed(2) + '\n'; });
console.log(out);
fs.writeFileSync('cashflow_008114_stopgain8.txt', out);
console.log('(已同时保存到 cashflow_008114_stopgain8.txt)');

File: verify_xirr.py
# 复刻 定投测算（净值）.html 的目标止盈定投现金流，并输出供 Excel 核对
import json, datetime

# ---------- 参数（请按你当时的设置修改） ----------
PARAMS = {
    'fundContains': '008114',   # 基金筛选
    'strategy': '5',            # 目标止盈定投
    'freq': 'monthly',
    'dayOfMonth': 'first',      # 每月首个交易日
    'baseAmount': 1000,
    'stopGainPct': 8,           # 止盈阈值 8%
    'stopGainSellRatio': 100,   # 赎回比例 100%
    'startDate': '2019-12-10',
    'endDate': '2026-07-07',
    'div': 'reinvest'           # 红利再投资
}

def parse(d): return datetime.datetime.strptime(d, '%Y-%m-%d')

# ---------- xirr（与 backtest.js 一致：dayFrac=(date-base)/365天，牛顿初值0.1，失败则二分） ----------
def xirr(cash_flows, flow_dates, guess=0.1):
    if len(cash_flows) != len(flow_dates) or len(cash_flows) < 2:
        return float('nan')
    paired = sorted(zip(flow_dates, cash_flows), key=lambda p: p[0])
    sdates = [p[0] for p in paired]
    sflows = [p[1] for p in paired]
    base = sdates[0]
    def dayfrac(j): return (sdates[j] - base).days / 365.0
    def npv(rate):
        return sum(sflows[j] / (1 + rate) ** dayfrac(j) for j in range(len(sflows)))
    tol = 1e-7
    rate = guess
    for _ in range(100):
        v = 0.0; d = 0.0
        for j in range(len(sflows)):
            f = dayfrac(j); term = (1 + rate) ** f
            v += sflows[j] / term
            d -= sflows[j] * f * (1 + rate) ** (f - 1)
        if abs(v) < tol: return rate
        if abs(d) < tol: break
        rate -= v / d
    lo, hi = -0.9999, 100.0
    fLo, fHi = npv(lo), npv(hi)
    if fLo * fHi > 0: return float('nan')
    for _ in range(200):
        mid = (lo + hi) / 2; fMid = npv(mid)
        if abs(fMid) < tol: return mid
        if fLo * fMid < 0: hi = mid
        else: lo = mid
    return (lo + hi) / 2

# ---------- 读取基金数据 ----------
raw = json.load(open('回测项目_2026-07-09.json', encoding='utf-8'))
funds = raw['fundsData']
fkey = [k for k in funds if PARAMS['fundContains'] in k][0]
fund = funds[fkey]
print('基金:', fkey, ' 交易日数:', len(fund['dates']))

st = parse(PARAMS['startDate']); en = parse(PARAMS['endDate'])
dates, date_strs, navs, divs = [], [], [], []
for i in range(len(fund['dates'])):
    d = parse(fund['dates'][i])
    if st <= d <= en:
        dates.append(d); date_strs.append(fund['dates'][i])
        navs.append(fund['nav'][i]); divs.append(fund['div'][i] or 0)
N = len(dates)
print('区间交易日数:', N)

dow = [d.weekday() for d in dates]            # Mon=0..Sun=6
dom = [int(s.split('-')[2]) for s in date_strs]

reinvest = PARAMS['div'] == 'reinvest'
strat = PARAMS['strategy']
is_invest = [False] * N
if strat == '1':
    is_invest[0] = True
else:
    freq = PARAMS['freq']
    if freq in ('weekly', 'biweekly'):
        wd = int(PARAMS.get('weekday', 1)) - 1   # 转成 Mon=0
        first = -1
        for k in range(N):
            if dow[k] == wd: first = k; break
        if first >= 0:
            for k in range(first, N):
                if dow[k] == wd:
                    if freq == 'biweekly':
                        if round((dates[k] - dates[first]).days / 7) % 2 != 0: continue
                    is_invest[k] = True
    else:
        if PARAMS['dayOfMonth'] == 'first':
            last = None
            for k in range(N):
                ym = date_strs[k][:7]
                if ym != last: is_invest[k] = True; last = ym
        else:
            domN = int(PARAMS['dayOfMonth'])
            for k in range(N):
                if dom[k] == domN: is_invest[k] = True

invest_cnt = sum(is_invest)
print('定投次数:', invest_cnt)

shares = 0.0; total_cash = 0.0; total_invested = 0.0; cost_basis = 0.0
stop_gain = 0; total_redeemed = 0.0; total_div_cash = 0.0
run_principal = 0.0; max_principal = 0.0; run_max_principal = 0.0
cash_flows, flow_dates = [], []

for k in range(N):
    nav = navs[k]; date = dates[k]
    if shares > 0 and divs[k] > 0:
        total_div = shares * divs[k]
        if reinvest: shares += total_div / nav
        else:
            total_cash += total_div; total_div_cash += total_div
            cash_flows.append(total_div); flow_dates.append(date)
    amt = 0.0
    if is_invest[k]:
        if strat in ('1', '2', '4', '5', '6'): amt = PARAMS['baseAmount']
        elif strat == '3':
            invest_cnt += 1; amt = invest_cnt * PARAMS['baseAmount'] - shares * nav
    if strat == '5':
        th = float(PARAMS['stopGainPct']) / 100
        sell_ratio = min(1.0, max(0.0, float(PARAMS['stopGainSellRatio']) / 100))
        round_principal = run_max_principal
        if round_principal > 0 and sell_ratio > 0:
            hold_mv = shares * nav
            if (hold_mv - cost_basis) / round_principal >= th:
                sell_shares = shares * sell_ratio
                proceeds = sell_shares * nav
                shares -= sell_shares; total_cash += proceeds; total_redeemed += proceeds
                cash_flows.append(proceeds); flow_dates.append(date)
                cost_basis *= (1 - sell_ratio)
                stop_gain += 1
                if sell_ratio >= 1 or shares < 1e-9:
                    run_principal = 0.0; run_max_principal = 0.0
    if amt != 0 and is_invest[k]:
        if amt > 0:
            shares += amt / nav; total_invested += amt; cost_basis += amt; run_principal += amt
            cash_flows.append(-amt); flow_dates.append(date)
        else:
            sell_shares = (-amt) / nav
            if sell_shares > shares: sell_shares = shares
            if sell_shares > 0:
                proceeds = sell_shares * nav
                shares -= sell_shares; total_cash += proceeds; total_redeemed += proceeds
                cash_flows.append(proceeds); flow_dates.append(date)
                denom = shares + sell_shares
                if denom > 0: cost_basis *= shares / denom
    run_max_principal = max(run_max_principal, run_principal)
    max_principal = max(max_principal, run_principal)

final_date = dates[N - 1]
cash_flows.append(shares * navs[N - 1]); flow_dates.append(final_date)

xirr_val = xirr(cash_flows, flow_dates)
final_asset = shares * navs[N - 1] + total_cash

print('\n===== 摘要 =====')
print('总投入本金 totalInvested     =', round(total_invested))
print('期末总资产 finalAsset        =', round(final_asset))
print('累积赎回 totalRedeemed       =', round(total_redeemed))
print('累计现金分红                 =', round(total_div_cash))
print('止盈次数 stopGainCount       =', stop_gain)
print('最大投入本金 maxPrincipal    =', round(max_principal))
print('期末剩余份额市值             =', round(shares * navs[N - 1]))
print('代码 xirr(牛顿 guess=0.1)    =', 'NaN' if xirr_val != xirr_val else f'{xirr_val*100:.4f}%')

# ---------- 扫描所有实根 ----------
def npv_rate(rate):
    paired = sorted(zip(flow_dates, cash_flows), key=lambda p: p[0])
    sd = [p[0] for p in paired]; sf = [p[1] for p in paired]
    bd = sd[0]
    return sum(sf[j] / (1 + rate) ** ((sd[j] - bd).days / 365.0) for j in range(len(sf)))

roots = []
prev_r = -0.9999; prev_v = npv_rate(prev_r)
r = -0.9999
while r <= 100.0:
    r += 0.0005
    v = npv_rate(r)
    if prev_v * v < 0:
        lo, hi = prev_r, r
        for _ in range(60):
            mid = (lo + hi) / 2
            if npv_rate(lo) * npv_rate(mid) <= 0: hi = mid
            else: lo = mid
        roots.append((lo + hi) / 2)
    prev_r, prev_v = r, v

print('\n===== 所有实根 (NPV=0 的 r) =====')
if not roots: print('无实根（区间内 NPV 不变号）')
else:
    for i, rt in enumerate(roots): print(f'根{i+1}: {rt*100:.4f}%')

# ---------- 现金流表（供 Excel 粘贴，日期<TAB>金额） ----------
print('\n===== 现金流表 (日期<TAB>金额) 复制粘贴到 Excel 两列 =====')
paired = sorted(zip(flow_dates, cash_flows), key=lambda p: p[0])
out = ''
for d, cf in paired:
    out += d.strftime('%Y-%m-%d') + '\t' + f'{cf:.2f}\n'
print(out)
open('cashflow_008114_stopgain8.txt', 'w', encoding='utf-8').write(out)
print('(已同时保存到 cashflow_008114_stopgain8.txt)')
